### ISSUE TITLE
Layout builder link issues

### ISSUE THREAD

**Issue by bspeare** [uiowa/uiowa#6730]
# Problem

The "Add content", "Delete component" and "Add component" links in layout builder have been producing unexpected results ever since the morning of the 9.5 update Tuesday, July 25 which may or may not be a coincidence :)  

- In Safari, https://iowajpec.org/node/2016/layout clicking the "Add content" link consistently reloads the page (for me) and the button is not usable.
  - From CLAS on 7/26 "I was going to add a new block on https://psychology.prod.drupal.uiowa.edu/research/areas-focus, but when clicking the “Add Component” it links me back to the Layout page and not bringing up the Blocks dropdown."
    -  I also had another user replicate the issue and they saw the same thing happening.
  - From Classrooms "Whenever I try to add a component to an existing section, it just reloads the page. If I try to add a section, it does so, but then when I try to add a component in that section it takes me to this page (in screenshot below)"
- Reported from CLAS, Tippie and Classrooms teams, the "Add component" and "Delete component" buttons have been taking users to the full page (see screenshots below)

![Screenshot 2023-08-01 at 10 14 06 AM](https://github.com/uiowa/uiowa/assets/1036433/e8e95aaa-9cd9-4dc5-bbb1-42fe90080354)

![Unknown-3](https://github.com/uiowa/uiowa/assets/1036433/a322b32f-5961-4522-a99f-3ff2e5edb93c)

# Steps to reproduce
Login to https://iowajpec.org/node/2016/layout and try clicking “Add component” in Safari. 

# Proposed solution
?

**Comment by bspeare** [comment]
Looks like a Cherwell ticket just came in describing the same bug.

**Comment by pyrello** [comment]
Not sure if this is related or not, but I'm seeing a CORS error related to SiteImprove in the console (I truncated the URL):
```
Access to XMLHttpRequest at 'https://id.siteimprove.com/connect/authorize... (redirected from 'https://my2.siteimprove.com/overlay/cms/cmssite?url=https://iowajpec.org/2023-mandela-washington-fellowship-copy-testing) from origin 'https://iowajpec.org' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Comment by pyrello** [comment]
I'm still seeing intermittent differences in behavior. Some "Add component" links seem to always trigger a reload, while others do only some of the time. I'm not noticing a clear pattern yet.

I tried uninstalling the SiteImprove module but that did not fix the issue.

**Comment by pyrello** [comment]
I'm also noticing that if I wait long enough (like while writing one of these comments) it appears that the behavior is loaded for every button and they work as expected.

**Comment by pyrello** [comment]
The pages where this behavior has been reported and we can recreate it (so far) seem to all have a lot of blocks on them. When I visit the layout tab for other pages on the same sites with less content, there is no issue at all. I'm wondering if this is a new form of a previous issue where we were bumping into the limit of how many blocks you could put on a page.

**Comment by pyrello** [comment]
Related: #6644

**Comment by pyrello** [comment]
Kind of related: https://www.drupal.org/project/layout_builder_restrictions/issues/3302842

> I did a little profiling and investigating and determined pretty definitively that this performance problem is due to the number of options that the interface has to *render* (in contrast to business logic or database calls). Specifically, it has to render a checkbox element for each of the blocks in the system *twice*, once for the Allow-list restrictions, and once for the deny-list restrictions.

This is in the Layout Builder Restrictions module. Its not the same issue that we are encountering, but it seems to suggest a similar problem, that we are encountering the limits of how many elements a browser can render.

**Comment by pyrello** [comment]
Summary of observations so far:
- Behavior on affected pages is intermittent (happens most of the time, but not every time) and the pattern of when it happens or not is not clear.
- Affected pages seem to be exclusively pages with a lot of content/blocks.
- Other pages on the same site with less content/blocks do not exhibit the behavior at all.
- I have not been able to reproduce locally.

Thoughts/next steps:
- While we seem to be getting a lot of reports all at once, I suspect that this is still only affecting a tiny percent of pages and users.
- We have an existing known issue that already covers what's happening: https://sitenow.uiowa.edu/known-issues#uiowa-issue-6644
- We may want to consider adding a warning when people exceed a certain number of blocks (although I'm not entirely clear what that limit should be or whether that is exactly the issue). We would reference the known issue in the message.
- I feel like this may be an issue where there are too many HTML elements being loaded on the page. I can't find any evidence to specifically back up this idea though.

**Comment by pyrello** [comment]
In trying to figure out how to resolve this issue, @bspeare switched LB Direct Add widget to the drop-button on the Iowa JPEC site and had James K. test it. There is a forwarded email in ITS-web that confirms that James had no problems with the drop-button widget. Ben is recommending that we switch the setting temporarily until we can resolve the underlying issue.

**Comment by quamsta** [comment]
It seems to be happening on this page I'm working on that doesn't have a lot of blocks:

https://labtanas.prod.drupal.uiowa.edu/

Dev Console does show this worrying Javascript error when clicking "Add component" and getting the refresh issue

<img width="936" alt="Screenshot 2023-08-02 at 11 03 14 AM" src="https://github.com/uiowa/uiowa/assets/472923/58114ff4-12a4-4b95-bd74-4e9755d3c01f">

**Comment by quamsta** [comment]
A couple more notes I've noticed that helps recreate the error:

- If I go to another site editor that's launched, aka a non .prod. domain and use Layout Builder, successfully click "Add Component" and go back to the .prod. site (like labtanas.prod.drupal.uiowa.edu), "Add Component" works just fine on the .prod. site until I hard refresh the .prod site and after hard refreshing, something seems to break.

- Also hard refreshing a site in my browser where "Add component" is actually working seems to break it.

**Comment by pyrello** [comment]
> It seems to be happening on this page I'm working on that doesn't have a lot of blocks:
>
> https://labtanas.prod.drupal.uiowa.edu/

@quamsta Is it the homepage? I can't get it to happen on that page even with a hard refresh.

**Comment by quamsta** [comment]
> > It seems to be happening on this page I'm working on that doesn't have a lot of blocks:
> > https://labtanas.prod.drupal.uiowa.edu/
> 
> @quamsta Is it the homepage? I can't get it to happen on that page even with a hard refresh.

Yeah, the homepage. It is really intermittent though, I can get the LB dialog to appear once in a great while, but really can't reliably reproduce the issue. :(

**Comment by quamsta** [comment]
@pyrello Oddly, one thing that does seem to reliably make the "Add Component" button work is just switching to a new browser I haven't even logged into the site with before. But after a bit of refreshing, all browsers seem to start exhibiting the same refresh bug. I started with Chrome, switched to Edge, then finally Firefox. All worked initially upon first using Layout Builder, but eventually none of them did.

**Comment by quamsta** [comment]
Also it seems to happen more when Chrome Dev Tools' Network Tab "Disable Cache" is checked, so I'm guessing there's a JS resource somewhere that might not be loading at the proper time to intercept the click event that's occurring on "Add Component," but it really is sort of unpredictable. It may be worth trying to recreate the issue while throttling your browser's speed to emulate slower connections? Not sure.

**Comment by quamsta** [comment]
Sorry to keep spamming this issue, but after blocking the network request for the very large FontAwesome all.min.js (449kb transferred), "Add Component" seems to more reliably work, with or without the cache disabled. Leading me to believe that the whatever's intercepting the click events for that button is possibly blocked by other large scripts loading first.

<img width="688" alt="Screenshot 2023-08-02 at 12 43 11 PM" src="https://github.com/uiowa/uiowa/assets/472923/548a10b5-9f3e-4a8e-8d1a-6d3abf23961f">

**Comment by bspeare** [comment]
Nice find @quamsta!

**Comment by quamsta** [comment]
Thanks @bspeare glad to help out. I will reiterate what I said in the Slack thread though, that removing a JS file from the stack is probably just a bandaid and this issue may be a symptom of some conflicting Javascript-related things happening and FontAwesome's role may just be in the realm of the straw-breaking-the-camel's-back-type thing.

I think in the long run it may be beneficial to do a bit of analysis on what scripts are loading when (which ones are deferred vs which ones are appended near the end of `</body>`, for example) and maybe adding some breakpoints or `console.log()` statements to the [lb-direct-add.js](https://git.drupalcode.org/project/lb_direct_add/-/blob/2.x/js/lb-direct-add.js) script to see exactly why it's not able to connect events to those buttons when requests are slower and/or the network is bogged down.

**Comment by quamsta** [comment]
https://github.com/uiowa/uiowa/pull/6741 Seems to have resolved this for me after some testing.
